Use new job search API for triggering Databricks job by name

alexott opened the pull request:
Existing implementation uses list API to find all possible jobs and filtering out them by name.  This leads to huge load on the API backend for workspaces with big number of jobs (> 1k).  Databricks Workflows team extended list API with search functionality.

ephraimbuddy commented on airflow/providers/databricks/hooks/databricks.py:
```diff
@@ -171,22 +171,41 @@ def list_jobs(self, limit: int = 25, offset: int = 0, expand_tasks: bool = False
         """
-        all_jobs = self.list_jobs()
-        matching_jobs = [j for j in all_jobs if j["settings"]["name"] == job_name]
+        if use_old_list_impl:  # For compatibility with old implementation
```
I wonder if it should be to `use_new_list_impl` unless this is for a major version bump

alexott replied:
I'll reimplement it a different way - that's why I converted it to draft.

alexott committed:
Use new job search API for triggering Databricks job by name

Existing implementation uses list API to find all possible jobs and filtering out them by
name. This leads to huge load on the API backend for workspaces with big number of jobs
(> 1k). Databricks Workflows team extended list API with search functionality.

ephraimbuddy commented on airflow/providers/databricks/hooks/databricks.py:
```diff
@@ -152,29 +152,38 @@ def submit_run(self, json: dict) -> int:
 
         while has_more:
-            json = {
+            jsn = {
```
How about we call this `json_data`? I feel it's better than an abbreviation. No strong opinion here

ephraimbuddy commented on airflow/providers/databricks/hooks/databricks.py:
```diff
@@ -152,29 +152,38 @@ def submit_run(self, json: dict) -> int:
+            if job_name:
+                jsn["name"] = job_name
+            response = self._do_api_call(LIST_JOBS_ENDPOINT, jsn)
+            jbs = response.get("jobs", [])
```
These abbreviations feel somehow, `databrick_jobs`...

alexott committed:
address review comments

alexott committed:
fix mypy error

alexott committed:
Merge branch 'main' into databricks-use-name-search-listing
